keras/backend/tensorflow_backend.py

```diff
@@ -3014,7 +3014,7 @@ def categorical_crossentropy(target, output, from_logits=False):
                                                        logits=output)
 
 
-def sparse_categorical_crossentropy(target, output, from_logits=False):
+def sparse_categorical_crossentropy(target, output, from_logits=False, data_format=None):
```

Comment:
@fchollet As I wrote yesterday, if I simply add an `axis` argument with default `-1`, then the whole flow leading up to the call to `sparse_categorical_crossentropy` needs to pass the axis, or it will simply default to `-1` and will continue to fail when the `image_data_format` is `'channels_first'`. The bug would not be solved at all.

There is currently nowhere in the flow that checks or passes the channels axis or the data format. To allow the user to set the channels axis directly without setting `K.image_data_format`, the channels axis argument would need to be added to the API and handled correctly throughout the training code, starting from the all functions called by the user (such as `fit` and `evaluate`), through the full training flow, and into the backend. This would require a far more major code restructuring.

In general, to fully support `'channels_first'` data format requires making changes in many places in the codebase, and is beyond my ability to fix by myself, which is why I opened issue #9670. In this pull request I'm taking a stab at fixing a few of the issues, but it's not a full solution. I know the other backends need to be fixed too, but I'm not familiar with them at all, which is why for now I made the fix only for TF and ran the test only with the TF backend. 

Passing an optional `data_format` may not be the ideal solution, but it does take at least one step in the right direction, and allows at least some of the functionality to work correctly with `'channels_first'`. At the same time it satisfies the requirements you wrote (that the data format can be explicitly passed by the user, and if not passed will default to `K.image_data_format`), and it is consistent with the behavior of many other functions in `tensorflow_backend`, which all do exactly that.

Reply:
@fchollet I never got an answer for this.

Reply:
@fchollet Please advise so I can finish this pull request.

Reply:
Please exclude changes related to `sparse_categorical_crossentropy` (or other losses) from this PR.

Reply:
@fchollet What do you mean? All the changes in this PR relate to `sparse_categorical_crossentropy`.

Like I said, these changes aren't a full fix to the problem of supporting `'channels_first'`, but they do solve one set of bugs and are consistent with the behavior elsewhere, so they seem to be a step in the right direction and should be accepted.

Reply:
@fchollet What's going on with this? Can you please approve this pull request. I can make the small fixes you mentioned to the documentation, but please let me know that the rest is fine before I put in any more work on this.

Reply:
@tiferet the API change in `sparse_categorical_crossentropy` is not something we can merge, sorry. Such an argument should be called `axis` and should default to `-1` (`K.image_data_format()` is a layer-level configuration argument and should not affect the default behavior of backend methods).

Reply:
@fchollet So what is the plan for addressing these issues? Will Keras simply not support ''channels_first''? Or do you intend to change the entire flow so that the channels axis is passed throughout, starting from the all functions called by the user (such as fit and evaluate), through the full training flow, and into the backend?

Reply:
For now, after adding `axis` in the crossentropy losses, you will have to use a different loss function when doing pixelwise classification (image segmentation) in NCHW:

```python
if K.data_format() == 'channels_first':
     loss = lambda y_true, y_pred: K.sparse_categorical_crossentropy(y_true, y_pred, axis=1)
else:
     loss = K.sparse_categorical_crossentropy
model.compile(optimizer=optimizer, loss=loss)
```

Reply:
@fchollet OK, so you do want me to change `sparse_categorical_crossentropy`, adding an `axis` argument with default `-1`, correct? Should I make that change and the other little fixes you pointed out and continue with the pull request?

Reply:
> so you do want me to change sparse_categorical_crossentropy, adding an axis argument with default -1, correct

You can do that, Yes. If you do that, please apply the change to all crossentropy methods (2 of them) and add that API keyword in all backends, since this is a backend-level change.

The other change in this PR is automated target shape inference when the user passes `sparse_categorical_crossentropy`. That change is OK.

Reply:
@fchollet I took care of all crossentropy losses in all backends. Note that `binary_crossentropy` does not need an `axis` argument, as its computation does not depend on the location of the channels axis. It works correctly with `channels_first` and `channels_last` as-is.

I added a test `test_model_with_crossentropy_losses_channels_first` that creates models with all three crossentropy losses and tests them with `channels_first` and `channels_last`, verifying that the losses come out the same with both `image_data_fromats`. I also made all the little corrections you asked for previously, so I think this PR should now be complete

Reply:
@fchollet Can you please check this PR? I've taken care of everything you asked for, and it should be ready to accept now.